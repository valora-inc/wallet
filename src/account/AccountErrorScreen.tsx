import React from 'react'
import { StyleSheet, Text } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import TextButton from 'src/components/TextButton'
import fontStyles from 'src/styles/fonts'
import { Spacing } from 'src/styles/styles'

interface Props {
  title: string
  testID: string
  description: string
  onPress(): void
  buttonLabel: string
  onPressSecondary?(): void
  secondaryButtonLabel?: string
}

function AccountErrorScreen({
  title,
  testID,
  description,
  onPress,
  buttonLabel,
  secondaryButtonLabel,
  onPressSecondary,
}: Props) {
  return (
    <SafeAreaView style={styles.content}>
      <Text style={styles.title} testID={testID}>
        {title}
      </Text>
      <Text style={styles.body}>{description}</Text>
      <TextButton onPress={onPress} testID={`${testID}Button`}>
        {buttonLabel}
      </TextButton>
      {secondaryButtonLabel && onPressSecondary && (
        <TextButton
          style={styles.secondaryButton}
          onPress={onPressSecondary}
          testID={`${testID}ButtonSecondary`}
        >
          {secondaryButtonLabel}
        </TextButton>
      )}
    </SafeAreaView>
  )
}

const styles = StyleSheet.create({
  content: {
    flex: 1,
    marginHorizontal: Spacing.Thick24,
    justifyContent: 'center',
    alignItems: 'center',
  },
  title: {
    ...fontStyles.h2,
    textAlign: 'center',
    paddingBottom: Spacing.Regular16,
    paddingTop: Spacing.Regular16,
  },
  body: {
    ...fontStyles.regular,
    textAlign: 'center',
    paddingBottom: Spacing.Thick24,
  },
  secondaryButton: {
    paddingTop: Spacing.Thick24,
  },
})

export default AccountErrorScreen
