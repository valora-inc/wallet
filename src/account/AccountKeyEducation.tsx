import { StackScreenProps, TransitionPresets } from '@react-navigation/stack'
import React, { useEffect } from 'react'
import { useTranslation } from 'react-i18next'
import Education, { EducationTopic, EmbeddedNavBar } from 'src/account/Education'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { BtnTypes } from 'src/components/Button'
import { accountKey1, accountKey2, accountKey3, accountKey4 } from 'src/images/Images'
import { noHeader } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { StackParamList } from 'src/navigator/types'

type Props = StackScreenProps<StackParamList, Screens.AccountKeyEducation>

export default function AccountKeyEducation(props: Props) {
  function onComplete() {
    ValoraAnalytics.track(OnboardingEvents.backup_education_complete)
    if (props.route.params?.nextScreen) {
      navigate(props.route.params?.nextScreen)
    } else {
      navigate(Screens.BackupPhrase)
    }
  }

  const { t } = useTranslation()

  const steps = useSteps()

  useEffect(() => {
    ValoraAnalytics.track(OnboardingEvents.backup_education_start)
  }, [])

  return (
    <Education
      embeddedNavBar={EmbeddedNavBar.Close}
      stepInfo={steps}
      onFinish={onComplete}
      finalButtonText={t('completeEducation')}
      buttonText={t('next')}
      finalButtonType={BtnTypes.PRIMARY}
    />
  )
}

AccountKeyEducation.navigationOptions = {
  ...noHeader,
  ...TransitionPresets.ModalTransition,
}

function useSteps() {
  const { t } = useTranslation()
  return React.useMemo(
    () =>
      [
        { image: accountKey1, topic: EducationTopic.backup },
        { image: accountKey2, topic: EducationTopic.backup },
        { image: accountKey3, topic: EducationTopic.backup },
        { image: accountKey4, topic: EducationTopic.backup },
      ].map((step, index) => {
        return {
          ...step,
          title: t(`guide.${index}.title`),
          text: t(`guide.${index}.text`),
        }
      }),
    []
  )
}
