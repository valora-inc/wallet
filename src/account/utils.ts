import { parsePhoneNumber } from '@celo/phone-utils'

const ADDRESS_LENGTH = 42 // TODO(ACT-1173): see if this can be replaced with a viem helper

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}

export function getPhoneNumberDetails(
  phoneNumber: string,
  countryCallingCode: string,
  countryCodeAlpha2: string
) {
  const phoneDetails = parsePhoneNumber(phoneNumber, countryCallingCode)

  if (phoneDetails) {
    return {
      // Show international display number to avoid confusion
      internationalPhoneNumber: phoneDetails.displayNumberInternational,
      e164Number: phoneDetails.e164Number,
      isValidNumber: true,
      countryCodeAlpha2: phoneDetails.regionCode!,
    }
  } else {
    return {
      internationalPhoneNumber: phoneNumber,
      e164Number: '',
      isValidNumber: false,
      countryCodeAlpha2,
    }
  }
}
