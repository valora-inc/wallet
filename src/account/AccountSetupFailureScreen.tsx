import React from 'react'
import { useTranslation } from 'react-i18next'
import RNExitApp from 'react-native-exit-app'
import AccountErrorScreen from 'src/account/AccountErrorScreen'
import { noHeaderGestureDisabled } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'

function AccounSetupFailureScreen() {
  const { t } = useTranslation()

  const onPressCloseApp = () => {
    RNExitApp.exitApp()
  }

  const onPressContactSupport = () => {
    navigate(Screens.SupportContact)
  }

  return (
    <AccountErrorScreen
      title={t('accountSetupFailed')}
      testID="AccountSetupFailure"
      description={t('accountSetupFailedDescription')}
      onPress={onPressCloseApp}
      buttonLabel={t('closeApp')}
      onPressSecondary={onPressContactSupport}
      secondaryButtonLabel={t('contactSupport')}
    />
  )
}

AccounSetupFailureScreen.navOptions = noHeaderGestureDisabled

export default AccounSetupFailureScreen
