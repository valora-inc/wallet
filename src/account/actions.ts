import { PincodeType } from 'src/account/reducer'

export enum Actions {
  CHOOSE_CREATE_ACCOUNT = 'ACCOUNT/CHOOSE_CREATE',
  CHOOSE_RESTORE_ACCOUNT = 'ACCOUNT/CHOOSE_RESTORE',
  START_STORE_WIPE_RECOVERY = 'ACCOUNT/START_STORE_WIPE_RECOVERY',
  CANCEL_CREATE_OR_RESTORE_ACCOUNT = 'ACCOUNT/CANCEL_CREATE_OR_RESTORE_ACCOUNT',
  SET_NAME = 'ACCOUNT/SET_NAME',
  SET_PHONE_NUMBER = 'ACCOUNT/SET_PHONE_NUMBER',
  SAVE_NAME = 'ACCOUNT/SAVE_NAME',
  DEV_MODE_TRIGGER_CLICKED = 'ACCOUNT/NAME_CLICKED',
  PHOTOSNUX_CLICKED = 'ACCOUNT/PHOTOSNUX_CLICKED',
  SET_PINCODE_SUCCESS = 'ACCOUNT/SET_PINCODE_SUCCESS',
  SET_PINCODE_FAILURE = 'ACCOUNT/SET_PINCODE_FAILURE',
  SET_ACCOUNT_CREATION_TIME = 'ACCOUNT/SET_ACCOUNT_CREATION_TIME',
  INITIALIZE_ACCOUNT = 'ACCOUNT/INITIALIZE_ACCOUNT',
  INITIALIZE_ACCOUNT_SUCCESS = 'ACCOUNT/INITIALIZE_ACCOUNT_SUCCESS',
  INITIALIZE_ACCOUNT_FAILURE = 'ACCOUNT/INITIALIZE_ACCOUNT_FAILURE',
  SET_BACKUP_COMPLETED = 'ACCOUNT/SET_BACKUP_COMPLETED',
  DISMISS_GET_VERIFIED = 'ACCOUNT/DISMISS_GET_VERIFIED',
  DISMISS_GOLD_EDUCATION = 'ACCOUNT/DISMISS_GOLD_EDUCATION',
  SET_USER_CONTACT_DETAILS = 'ACCOUNT/SET_USER_CONTACT_DETAILS',
  ACCEPT_TERMS = 'ACCOUNT/ACCEPT_TERMS',
  CLEAR_STORED_ACCOUNT = 'ACCOUNT/CLEAR_STORED_ACCOUNT',
  PROFILE_UPLOADED = 'ACCOUNT/PROFILE_UPLOADED',
  SAVE_SIGNED_MESSAGE = 'ACCOUNT/SAVE_SIGNED_MESSAGE',
  SET_CELO_EDUCATION_COMPLETED = 'ACCOUNT/SET_CELO_EDUCATION_COMPLETED',
  RECOVERY_PHRASE_IN_ONBOARDING_STARTED = 'ACCOUNT/RECOVERY_PHRASE_IN_ONBOARDING_STARTED',
  RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED = 'ACCOUNT/RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED',
}

export interface ChooseCreateAccountAction {
  type: Actions.CHOOSE_CREATE_ACCOUNT
  now: number
}
interface ChooseRestoreAccountAction {
  type: Actions.CHOOSE_RESTORE_ACCOUNT
}

interface StartStoreWipeRecoveryAction {
  type: Actions.START_STORE_WIPE_RECOVERY
  accountToRecover: string
}

interface CancelCreateOrRestoreAccountAction {
  type: Actions.CANCEL_CREATE_OR_RESTORE_ACCOUNT
}

interface SetNameAction {
  type: Actions.SET_NAME
  name: string
}

interface SetPhoneNumberAction {
  type: Actions.SET_PHONE_NUMBER
  e164PhoneNumber: string
  countryCode: string
}

interface SaveNameAction {
  type: Actions.SAVE_NAME
  name: string
}
interface DevModeTriggerClickedAction {
  type: Actions.DEV_MODE_TRIGGER_CLICKED
}

interface AcceptTermsAction {
  type: Actions.ACCEPT_TERMS
}

interface PhotosNUXClickedAction {
  type: Actions.PHOTOSNUX_CLICKED
}

interface SetPincodeSuccessAction {
  type: Actions.SET_PINCODE_SUCCESS
  pincodeType: PincodeType
}

interface SetPincodeFailureAction {
  type: Actions.SET_PINCODE_FAILURE
}

interface InitializeAccountAction {
  type: Actions.INITIALIZE_ACCOUNT
}

interface InitializeAccountSuccessAction {
  type: Actions.INITIALIZE_ACCOUNT_SUCCESS
}

interface SetAccountCreationAction {
  type: Actions.SET_ACCOUNT_CREATION_TIME
  now: number
}

interface SetBackupCompletedAction {
  type: Actions.SET_BACKUP_COMPLETED
}

interface DismissGetVerifiedAction {
  type: Actions.DISMISS_GET_VERIFIED
}

interface DismissGoldEducationAction {
  type: Actions.DISMISS_GOLD_EDUCATION
}

interface SetContactDetailsAction {
  type: Actions.SET_USER_CONTACT_DETAILS
  contactId: string
  thumbnailPath: string | null
}

export interface ClearStoredAccountAction {
  type: Actions.CLEAR_STORED_ACCOUNT
  account: string
}

interface ProfileUploadedAction {
  type: Actions.PROFILE_UPLOADED
}

interface SaveSignedMessage {
  type: Actions.SAVE_SIGNED_MESSAGE
}

interface SetCeloEducationCompletedAction {
  type: Actions.SET_CELO_EDUCATION_COMPLETED
  celoEducationCompleted: boolean
}

interface RecoveryPhraseInOnboardingStarted {
  type: Actions.RECOVERY_PHRASE_IN_ONBOARDING_STARTED
}

interface RecoveryPhraseInOnboardingCompleted {
  type: Actions.RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED
}

export type ActionTypes =
  | ChooseCreateAccountAction
  | ChooseRestoreAccountAction
  | StartStoreWipeRecoveryAction
  | CancelCreateOrRestoreAccountAction
  | SetNameAction
  | SetPhoneNumberAction
  | SaveNameAction
  | DevModeTriggerClickedAction
  | PhotosNUXClickedAction
  | SetPincodeSuccessAction
  | SetPincodeFailureAction
  | InitializeAccountAction
  | InitializeAccountSuccessAction
  | SetAccountCreationAction
  | SetBackupCompletedAction
  | DismissGetVerifiedAction
  | DismissGoldEducationAction
  | SetContactDetailsAction
  | AcceptTermsAction
  | ClearStoredAccountAction
  | ProfileUploadedAction
  | SaveSignedMessage
  | SetCeloEducationCompletedAction
  | RecoveryPhraseInOnboardingStarted
  | RecoveryPhraseInOnboardingCompleted

export function chooseCreateAccount(now: number): ChooseCreateAccountAction {
  return {
    type: Actions.CHOOSE_CREATE_ACCOUNT,
    now,
  }
}

export function chooseRestoreAccount(): ChooseRestoreAccountAction {
  return {
    type: Actions.CHOOSE_RESTORE_ACCOUNT,
  }
}

export function startStoreWipeRecovery(accountToRecover: string): StartStoreWipeRecoveryAction {
  return {
    type: Actions.START_STORE_WIPE_RECOVERY,
    accountToRecover,
  }
}

export function cancelCreateOrRestoreAccount(): CancelCreateOrRestoreAccountAction {
  return {
    type: Actions.CANCEL_CREATE_OR_RESTORE_ACCOUNT,
  }
}

export function acceptTerms(): AcceptTermsAction {
  return {
    type: Actions.ACCEPT_TERMS,
  }
}

export function saveName(name: string): SaveNameAction {
  return {
    type: Actions.SAVE_NAME,
    name,
  }
}
export const devModeTriggerClicked = (): DevModeTriggerClickedAction => ({
  type: Actions.DEV_MODE_TRIGGER_CLICKED,
})

export const setPincodeSuccess = (pincodeType: PincodeType): SetPincodeSuccessAction => ({
  type: Actions.SET_PINCODE_SUCCESS,
  pincodeType,
})

export const initializeAccount = (): InitializeAccountAction => ({
  type: Actions.INITIALIZE_ACCOUNT,
})

export const initializeAccountSuccess = (): InitializeAccountSuccessAction => ({
  type: Actions.INITIALIZE_ACCOUNT_SUCCESS,
})

export const setAccountCreationTime = (now: number): SetAccountCreationAction => ({
  type: Actions.SET_ACCOUNT_CREATION_TIME,
  now,
})

export const setBackupCompleted = (): SetBackupCompletedAction => ({
  type: Actions.SET_BACKUP_COMPLETED,
})

export const dismissGetVerified = (): DismissGetVerifiedAction => ({
  type: Actions.DISMISS_GET_VERIFIED,
})

export const dismissGoldEducation = (): DismissGoldEducationAction => ({
  type: Actions.DISMISS_GOLD_EDUCATION,
})

export const setUserContactDetails = (
  contactId: string,
  thumbnailPath: string | null
): SetContactDetailsAction => ({
  type: Actions.SET_USER_CONTACT_DETAILS,
  contactId,
  thumbnailPath,
})

export const clearStoredAccount = (account: string): ClearStoredAccountAction => ({
  type: Actions.CLEAR_STORED_ACCOUNT,
  account,
})

export const saveSignedMessage = (): SaveSignedMessage => ({
  type: Actions.SAVE_SIGNED_MESSAGE,
})

export const setGoldEducationCompleted = (): SetCeloEducationCompletedAction => ({
  type: Actions.SET_CELO_EDUCATION_COMPLETED,
  celoEducationCompleted: true,
})

export const recoveryPhraseInOnboardingStarted = (): RecoveryPhraseInOnboardingStarted => ({
  type: Actions.RECOVERY_PHRASE_IN_ONBOARDING_STARTED,
})

export const recoveryPhraseInOnboardingCompleted = (): RecoveryPhraseInOnboardingCompleted => ({
  type: Actions.RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED,
})
