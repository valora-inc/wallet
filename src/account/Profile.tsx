import { NativeStackScreenProps } from '@react-navigation/native-stack'
import React, { useMemo, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, View } from 'react-native'
import * as RNFS from 'react-native-fs'
import { SafeAreaView } from 'react-native-safe-area-context'
import { useDispatch, useSelector } from 'react-redux'
import { saveNameAndPicture } from 'src/account/actions'
import { nameSelector, pictureSelector } from 'src/account/selectors'
import { showError, showMessage } from 'src/alert/actions'
import { SettingsEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import Button from 'src/components/Button'
import CancelButton from 'src/components/CancelButton'
import KeyboardAwareScrollView from 'src/components/KeyboardAwareScrollView'
import KeyboardSpacer from 'src/components/KeyboardSpacer'
import TextInput from 'src/components/TextInput'
import Touchable from 'src/components/Touchable'
import i18n from 'src/i18n'
import { generateRandomUsername } from 'src/nameGenerator'
import { emptyHeader } from 'src/navigator/Headers'
import { Screens } from 'src/navigator/Screens'
import { StackParamList } from 'src/navigator/types'
import PictureInput from 'src/onboarding/registration/PictureInput'
import colors from 'src/styles/colors'
import fontStyles from 'src/styles/fonts'
import { saveProfilePicture } from 'src/utils/image'
import Logger from 'src/utils/Logger'

type Props = NativeStackScreenProps<StackParamList, Screens.Profile>

function Profile({ navigation, route }: Props) {
  const { t } = useTranslation()
  const [newName, setNewName] = useState(useSelector(nameSelector) ?? '')
  const picturePath = useSelector(pictureSelector)
  const [newPictureUri, setNewPictureUri] = useState(picturePath || null)
  const exampleName = useMemo(() => generateRandomUsername(), [])

  const dispatch = useDispatch()

  const onSave = () => {
    if (newName.length === 0) {
      dispatch(showError(ErrorMessages.MISSING_FULL_NAME))
      return
    }
    ValoraAnalytics.track(SettingsEvents.profile_save)
    dispatch(saveNameAndPicture(newName, newPictureUri))
    dispatch(showMessage(t('namePictureSaved')))
    navigation.goBack()

    // Delete old profile pictures if necessary.
    if (picturePath && picturePath !== newPictureUri) {
      RNFS.unlink(picturePath).catch((e) => {
        Logger.error('Profile', 'Error deleting old profile picture:', e)
      })
    }
  }

  const onPictureChosen = async (pictureDataUrl: string | null) => {
    if (!pictureDataUrl) {
      setNewPictureUri(null)
      ValoraAnalytics.track(SettingsEvents.profile_photo_removed)
    } else {
      try {
        const newPicturePath = await saveProfilePicture(pictureDataUrl)
        setNewPictureUri(newPicturePath)
        ValoraAnalytics.track(SettingsEvents.profile_photo_chosen)
      } catch (error) {
        dispatch(showError(ErrorMessages.PICTURE_LOAD_FAILED))
      }
    }
  }

  const updateName = (updatedName: string) => {
    setNewName(updatedName)
  }

  const generateName = () => {
    ValoraAnalytics.track(SettingsEvents.profile_generate_name)
    updateName(generateRandomUsername())
  }

  return (
    <SafeAreaView style={styles.container} edges={['bottom']}>
      <KeyboardAwareScrollView testID="ProfileScrollView">
        <View style={styles.accountProfile}>
          <PictureInput picture={newPictureUri} onPhotoChosen={onPictureChosen} />
        </View>
        <View style={styles.inputContainer}>
          <TextInput
            placeholder={t('profileScreen.namePlaceholder', { exampleName }) ?? undefined}
            testID="ProfileEditName"
            onChangeText={updateName}
            value={newName ?? t('unknown')}
          />
        </View>
        <View style={styles.ctaContainer}>
          <Button
            style={styles.generateButton}
            text={t('profileScreen.generateName')}
            onPress={generateName}
            testID="GenerateNameButton"
          />
          <Touchable testID="SaveButton" onPress={onSave}>
            <Text style={styles.saveButton}>{t('save')}</Text>
          </Touchable>
        </View>
      </KeyboardAwareScrollView>
      <KeyboardSpacer />
    </SafeAreaView>
  )
}

Profile.navigationOptions = ({ navigation }: Props) => {
  const onCancel = () => {
    ValoraAnalytics.track(SettingsEvents.profile_cancel)
    navigation.goBack()
  }
  return {
    ...emptyHeader,
    headerTitle: i18n.t('editProfile'),
    headerLeft: () => <CancelButton onCancel={onCancel} />,
  }
}

export default Profile

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  accountProfile: {
    paddingLeft: 10,
    paddingTop: 24,
    paddingRight: 15,
    paddingBottom: 37,
    flexDirection: 'column',
    alignItems: 'center',
  },
  inputContainer: {
    borderBottomWidth: 1,
    borderBottomColor: colors.gray2,
    marginHorizontal: 24,
  },
  ctaContainer: {
    alignItems: 'center',
  },
  generateButton: {
    marginVertical: 24,
  },
  saveButton: {
    ...fontStyles.regular,
    color: colors.primary,
  },
})
