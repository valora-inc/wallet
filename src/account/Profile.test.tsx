import { fireEvent, render } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import { MockStoreEnhanced } from 'redux-mock-store'
import { saveNameAndPicture } from 'src/account/actions'
import Profile from 'src/account/Profile'
import { showError, showMessage } from 'src/alert/actions'
import { SettingsEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { generateRandomUsername } from 'src/nameGenerator'
import { Screens } from 'src/navigator/Screens'
import { RootState } from 'src/redux/reducers'
import { createMockStore, getMockStackScreenProps } from 'test/utils'

jest.mock('src/analytics/ValoraAnalytics')
jest.mock('src/nameGenerator')

describe('Profile', () => {
  let store: MockStoreEnhanced
  beforeEach(() => {
    jest.clearAllMocks()
    store = createMockStore({})
  })

  it('renders correctly', () => {
    const { toJSON } = render(
      <Provider store={store}>
        <Profile {...getMockStackScreenProps(Screens.Profile)} />
      </Provider>
    )
    expect(toJSON()).toMatchSnapshot()
  })

  describe('when updating name', () => {
    const name = 'New Name'
    it('edits name', () => {
      const { getByDisplayValue, getByTestId } = render(
        <Provider store={store}>
          <Profile {...getMockStackScreenProps(Screens.Profile)} />
        </Provider>
      )

      const input = getByDisplayValue((store.getState() as RootState).account.name!)
      fireEvent.changeText(input, name)
      expect(store.getActions().length).toEqual(0)

      fireEvent.press(getByTestId('SaveButton'))
      expect(store.getActions()).toEqual([
        saveNameAndPicture(name, null),
        showMessage('namePictureSaved'),
      ])
      expect(ValoraAnalytics.track).toHaveBeenCalledWith(SettingsEvents.profile_save)
    })

    it('generate name fills in name with random name', () => {
      jest.mocked(generateRandomUsername).mockReturnValue('Random username')
      const { getByTestId } = render(
        <Provider store={store}>
          <Profile {...getMockStackScreenProps(Screens.Profile)} />
        </Provider>
      )
      fireEvent.press(getByTestId('GenerateNameButton'))
      expect(store.getActions().length).toEqual(0)

      fireEvent.press(getByTestId('SaveButton'))
      expect(store.getActions()).toEqual([
        saveNameAndPicture('Random username', null),
        showMessage('namePictureSaved'),
      ])
      expect(ValoraAnalytics.track).toHaveBeenCalledWith(SettingsEvents.profile_generate_name)
      expect(ValoraAnalytics.track).toHaveBeenCalledWith(SettingsEvents.profile_save)
    })

    it('serves error banner when attempting to save empty name', () => {
      const { getByDisplayValue, getByTestId } = render(
        <Provider store={store}>
          <Profile {...getMockStackScreenProps(Screens.Profile)} />
        </Provider>
      )

      const input = getByDisplayValue((store.getState() as RootState).account.name!)
      fireEvent.changeText(input, '')

      fireEvent.press(getByTestId('SaveButton'))

      expect(store.getActions()).toEqual([showError(ErrorMessages.MISSING_FULL_NAME)])
      expect(ValoraAnalytics.track).not.toHaveBeenCalledWith(SettingsEvents.profile_save)
    })
  })
})
