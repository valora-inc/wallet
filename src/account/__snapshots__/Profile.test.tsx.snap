// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "off",
      "right": "off",
      "top": "off",
    }
  }
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    keyboardShouldPersistTaps="handled"
    testID="ProfileScrollView"
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "column",
            "paddingBottom": 37,
            "paddingLeft": 10,
            "paddingRight": 15,
            "paddingTop": 24,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          focusable={true}
          nativeBackgroundAndroid={
            {
              "attribute": "selectableItemBackground",
              "rippleRadius": undefined,
              "type": "ThemeAttrAndroid",
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "alignSelf": "center",
                "borderRadius": 32,
                "height": 64,
                "justifyContent": "center",
                "width": 64,
              },
            ]
          }
          testID="PictureInput"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "justifyContent": "center",
                  },
                  {
                    "backgroundColor": "hsl(0, 53%, 93%)",
                    "borderRadius": 32,
                    "height": 64,
                    "width": 64,
                  },
                ]
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  [
                    {
                      "color": "#FFFFFF",
                      "fontFamily": "Inter-Medium",
                      "fontSize": 16,
                    },
                    {
                      "color": "hsl(0, 67%, 24%)",
                      "fontSize": 32,
                    },
                  ]
                }
              >
                J
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#F8F9F9",
                "borderRadius": 20,
                "bottom": -13,
                "height": 40,
                "justifyContent": "center",
                "left": 37,
                "position": "absolute",
                "width": 40,
              }
            }
          >
            <svg
              fill="none"
              height={13}
              style={{}}
              viewBox="0 0 13 13"
              width={13}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M10.8667 4.94998L8.03333 2.14998L8.96667 1.21665C9.22222 0.961091 9.53622 0.833313 9.90867 0.833313C10.2807 0.833313 10.5944 0.961091 10.85 1.21665L11.7833 2.14998C12.0389 2.40554 12.1722 2.71398 12.1833 3.07531C12.1944 3.4362 12.0722 3.74442 11.8167 3.99998L10.8667 4.94998ZM9.9 5.93331L2.83333 13H0V10.1666L7.06667 3.09998L9.9 5.93331Z"
                fill="#178154"
                fill-rule="evenodd"
                style={{}}
              />
            </svg>
          </View>
        </View>
        <Modal
          animationType="none"
          deviceHeight={1334}
          deviceWidth={null}
          hardwareAccelerated={false}
          hideModalContentWhileAnimating={false}
          onBackdropPress={[Function]}
          onModalHide={[Function]}
          onModalWillHide={[Function]}
          onModalWillShow={[Function]}
          onRequestClose={[Function]}
          panResponderThreshold={4}
          scrollHorizontal={false}
          scrollOffset={0}
          scrollOffsetMax={0}
          scrollTo={null}
          statusBarTranslucent={true}
          supportedOrientations={
            [
              "portrait",
              "landscape",
            ]
          }
          swipeThreshold={100}
          transparent={true}
          visible={false}
        />
      </View>
      <View
        style={
          {
            "borderBottomColor": "#E6E6E6",
            "borderBottomWidth": 1,
            "marginHorizontal": 24,
          }
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <TextInput
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="profileScreen.namePlaceholder, {}"
            style={
              [
                {
                  "color": "#2E3338",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
                {
                  "flex": 1,
                  "lineHeight": 22,
                  "paddingHorizontal": 0,
                  "paddingVertical": 12,
                },
                undefined,
              ]
            }
            testID="ProfileEditName"
            value="John Doe"
          />
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
          }
        }
      >
        <View
          style={
            [
              {
                "flexDirection": "row",
              },
              {
                "marginVertical": 24,
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 100,
                },
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              focusable={true}
              nativeBackgroundAndroid={
                {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "#1AB775",
                    "flexDirection": "row",
                    "height": 48,
                    "justifyContent": "center",
                    "minWidth": 120,
                    "opacity": undefined,
                    "paddingHorizontal": 24,
                    "paddingVertical": 5,
                  },
                  undefined,
                ]
              }
              testID="GenerateNameButton"
            >
              <Text
                maxFontSizeMultiplier={1}
                style={
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "Inter-SemiBold",
                    "fontSize": 16,
                    "lineHeight": 22,
                    "marginLeft": 0,
                    "marginRight": 0,
                  }
                }
              >
                profileScreen.generateName
              </Text>
            </View>
          </View>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          focusable={true}
          nativeBackgroundAndroid={
            {
              "attribute": "selectableItemBackground",
              "rippleRadius": undefined,
              "type": "ThemeAttrAndroid",
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="SaveButton"
        >
          <Text
            style={
              {
                "color": "#1AB775",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
              }
            }
          >
            save
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    collapsable={false}
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "right": 0,
        },
        {
          "height": 0,
        },
        undefined,
      ]
    }
  />
</RNCSafeAreaView>
`;
